### ISSUE TITLE
Reassigning variable changes type in preceding code

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#34076]
Given the code on the left, the type of `token` is narrowed from `Token` to `IdentifierToken`.

However, if I uncomment the line that reassigns it further down, it's type changes back to `Token` for the lines above. I would've expected it only to affect code that came after the reassignment.

![tokens](https://user-images.githubusercontent.com/1078012/43678307-1d78f552-9809-11e8-8f67-ff5c19d1229c.png)

**Comment by eernstg** [comment]
This is working as intended: The specification says (section 19.1.1 Type Promotion) that 'We say that a variable v is known to have type T whenever we allow the type of v to be promoted. The exact circumstances when type promotion is allowed are given in the relevant sections of the specification (16.23, 16.21 and 17.5).'

If you check those three locations then it always says that v is not potentially mutated.

So what you've witnessed is the situation where the variable 'token' is promoted to have type `IdentifierToken` vs. the situation where it is not promoted, and it's the assignment that makes the difference. The analysis is not flow sensitive, that is, it makes no difference where the assignment occurs.

Basically, in order to obtain promotion you should treat the variable as final.

We could easily have _required_ any promoted variable to be final, which would have removed this (admittedly real) source of confusion. However, that rule would have introduced a strict requirement for a lot of occurrences of `final` on local variable and parameter declarations, and that was considered too inconvenient in practice.

**Comment by eernstg** [comment]
Thanks! I hope that we can get a lightweight syntax for final variables (`x := e` would declare `x` as a final local variable with the static type of `e`). It might then be easier to maintain the rule (e.g., via a lint) that only final variables should be promoted.
